### ISSUE TITLE
Tags of influx queries should not be saved in a fields name

### ISSUE THREAD

**Issue by MariusBrill** [grafana/grafana#30498]
**What would you like to be added**:
Currently, whenever I create a flux query, fields are returned with all the corresponding tags in their names: 
![grafana_table](https://user-images.githubusercontent.com/53812669/105355302-3b4c4700-5bf2-11eb-9ace-1776327761e8.PNG). 

The preferred behavior would be similar to the behavior of the influx Data Explorer, where each tag is a field in a row:
![influx_table](https://user-images.githubusercontent.com/53812669/105355354-515a0780-5bf2-11eb-8355-7e21ab39287d.PNG)
.

**Why is this needed**:
As soon as i use the repeat function of a row, i cannot use any transform statements since each value's name is unique due to the tags being part of the value names. Therefore it is impossible to e.g. rename usage_percentage {........} to just just "Storage Used in %" which would be possible if the field was just named usage_percentage with alle the tags contained in other fields of one row.

**Comment by rovo89** [comment]
I use the "Labels to fields" transformation for this, it's working quite well.

**Comment by MariusBrill** [comment]
@rovo89  Thanks this indeed does the trick! =) 

However i still think this could be a default behavior for these queries.

**Comment by aocenas** [comment]
@MariusBrill I think the `Labels to fields` transform is the way to go here. The problem is in grafana labels on fields is a generic concept and not sure if it always makes sense to spread it out. For example labels are great if you have 2 `usage_percentage` fields that are differentiated by the labels. Spreading them out wouldn't actually help cause you will loose the association of what labels belong to which value column.

If you think there is some specific situation or query where this should actually happen, let us know.

**Comment by philomory** [comment]
@aocenas Isn't the current behavior just, always wrong, as well, though? Whether making `Labels to fields` on by default is reasonable or not, I have no idea, but with the current implementation, Grafana is basically taking the string representation of the first row of the results, and using it as the name of the `_value` column for all rows. Shouldn't it just say `_value` (or `Value`, since `_time` is transformed to `Time`)?

**Comment by quentinmit** [comment]
Because the fields are all getting reported as labels, there is also the problem that they're all getting converted to strings. AFAICT there's no way to turn them back into numbers in Grafana. This means it's impossible for me to get latitudes and longitudes out of Influx that can be plotted with the Worldmap plugin, for instance.

Also, when I use "Labels to fields" it seems like that drops the original value from the _value field? I'm not quite sure, though - maybe I'm just reading the transformation debug view wrong.

**Comment by aocenas** [comment]
@philomory can you share the query you are using?

**Comment by philomory** [comment]
@aocenas I didn't keep the original query around that I used for that screenshot, but I can get essentially the same results with this (somewhat pointless) query:

```
from(bucket: "live")
  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r._measurement == "rabbitmq_node" and r._field == "uptime")
  |> filter(fn: (r) => r.environment == "${environment}" and r.mq_cluster_name == "${cluster}")
  |> group(columns: ["host"])
```

In this query, each table in the stream (which Grafana transforms into a data frame) represents a distinct host in the RabbitMQ cluster. The "group key" is literally `["host"]`; each table in the stream has a different value for the "host" key, and all rows in a given table have the same host. And, the drop-down at the bottom of the table correctly switches between the three (in my case) different hosts in the selected cluster. But, instead of having the value of the host column displayed in the drop down for each row, each row just says `rabbitmq_node`, as you can see in the screenshot above. Conversely, one you _pick_ a table, the "value" column, which should _probably_ say "uptime" (or just "value"), instead looks like this:

![image](https://user-images.githubusercontent.com/200123/106504998-c982c180-646b-11eb-8d6e-4deb77abb650.png)

This is... fine. It does at least say "uptime" in there, and, nothing in it is technically inaccurate... for _this_ query. But the title bar looks the same in this query:

```
from(bucket: "live")
  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r._measurement == "rabbitmq_node" and r._field == "uptime")
  |> filter(fn: (r) => r.environment == "${environment}" and r.mq_cluster_name == "${cluster}")
  |> group()
```

In this query, all the records for the different hosts are grouped into a single table, but, the header of the value column still says `host=mq-1.test.mmem.co`.

**Comment by hugohaggmark** [comment]
@grafana/observability-squad I'll remove the needs more info label on this one and add needs investigation.

**Comment by philomory** [comment]
It's difficult for me to say for certain, because I'm not _super_ familiar with Grafana internals, but, I think that the problem here has multiple facets:

* First, non-value, non-time columns like `host`, `mq_cluster_name`, etc, are all treated the same (they become labels), regardless of whether they are part of the Group Key or not.
* Second, the only fields in the data frames are "Time" (from the `_time` column) and "Uptime" (which gets its data from the `_value` column and its name from the `_field` column. All the labels the labels are then attached to the "Uptime" field.
* Because the labels are attached to a _field_ and not a _row_, each label can only take one value per data frame. Each label appears to take the value of the corresponding column in the first record in that table.
* The frame name, meanwhile, is taken from the `_measurement` column. This is almost always going to be wrong.

The way this _should_ probably work, is roughly as follows:
* Each `FluxColumn` record in the array returned by the `FluxTableMetadata.Columns()` has a property, `group`, which can be checked via `IsGroup()`.
* The list of columns in the frame should be split up into those that are group keys, and those that are not.
* Any column that is a group key should be used to construct the _name_ of that Data Frame, the values of all of the group keys, as a set, define the identity of that frame.
* Any column that is _not_ a group key should almost certainly become a field, because each row in the table can and probably will have a separate value for that column.
* For the special case of the columns `_field` and `_value`, assuming `_value` isn't a group key (which it usually won't be) and `_field` is either a group key or, at least, has the same value for every row in this data frame, then, the `_value` column can be renamed the value of the `_field` column, and the `_field` column left out of the data frame.
* There shouldn't be any field-labels at all.

That's my best guess, anyway, based on what I see from looking at the code, looking at the frame data from sample queries, and reading the documentation on Data Frames.

**Comment by gabor** [comment]
when https://github.com/grafana/grafana/issues/32842 happens it should help with this situation, maybe even fix it. either way, we can come back and re-evaluate this issue when the mentioned issue is done.

**Comment by gabor** [comment]
we changed how complex flux queries are parsed. it is available in 8.0.0-beta1 or later. it should fix the problem ("tag" columns and "data" columns are parsed more correctly now and no data-column is ignored), but the response format might be slightly different because of this.

if it does not solve your issue completely, please open a new issue (i'd like to not mix old-behavior and new-behavior in the same issue, to keep issues easy to understand).
